src/main/java/com/minecolonies/coremod/entity/citizen/citizenhandlers/CitizenSkillHandler.java

```diff
@@ -0,0 +1,252 @@
+
+        @NotNull final ListNBT levelTagList = new ListNBT();
+        for (@NotNull final Map.Entry<Skill, Tuple<Integer, Double>> entry : skillMap.entrySet())
+        {
+            @NotNull final CompoundNBT levelCompound = new CompoundNBT();
+            levelCompound.putInt(TAG_SKILL, entry.getKey().ordinal());
+            levelCompound.putInt(TAG_LEVEL, entry.getValue().getA());
+            levelCompound.putDouble(TAG_EXPERIENCE, entry.getValue().getB());
+            levelTagList.add(levelCompound);
+        }
+        compoundNBT.put(TAG_LEVEL_MAP, levelTagList);
+
+        return compoundNBT;
+    }
+
+    @Override
+    public void read(@NotNull final CompoundNBT compoundNBT)
+    {
+        final ListNBT levelTagList = compoundNBT.getList(TAG_LEVEL_MAP, Constants.NBT.TAG_COMPOUND);
+        for (int i = 0; i < levelTagList.size(); ++i)
+        {
+            final CompoundNBT levelExperienceAtJob = levelTagList.getCompound(i);
+            skillMap.put(Skill.values()[levelExperienceAtJob.getInt(TAG_SKILL)],
+              new Tuple<>(Math.min(levelExperienceAtJob.getInt(TAG_LEVEL), MAX_CITIZEN_LEVEL), levelExperienceAtJob.getDouble(TAG_EXPERIENCE)));
+        }
+    }
+
+    @Override
+    public void tryLevelUpIntelligence(@NotNull final Random random, final int customChance, @NotNull final ICitizenData citizen)
+    {
+        if ((customChance > 0 && random.nextInt(customChance) > 0) || (customChance < 1 && random.nextInt(CHANCE_TO_LEVEL) > 0))
+        {
+            return;
+        }
+
+        final int levelCap = (int) citizen.getCitizenHappinessHandler().getHappiness();
+        if (skillMap.get(Skill.Intelligence).getB() < levelCap * 9)
+        {
+            addXpToSkill(Skill.Intelligence, 10, citizen);
+        }
+    }
+
+    @Override
+    public int getLevel(@NotNull final Skill skill)
+    {
+        return skillMap.get(skill).getA();
+    }
+
+    @Override
+    public void incrementLevel(@NotNull final Skill skill, final int level)
+    {
+        final Tuple<Integer, Double> current = skillMap.get(skill);
+        skillMap.put(skill, new Tuple<>(Math.min(MAX_CITIZEN_LEVEL, Math.max(current.getA() + level, 1)), current.getB()));
+    }
+
+    @Override
+    public void addXpToSkill(final Skill skill, final double xp, final ICitizenData data)
+    {
+        final Tuple<Integer, Double> tuple = skillMap.get(skill);
+        int level = tuple.getA();
+        final double currentXp = tuple.getB();
+
+        final IBuilding home = data.getHomeBuilding();
+
+        final double citizenHutLevel = home == null ? 0 : home.getBuildingLevel();
+        final double citizenHutMaxLevel = home == null ? 5 : home.getMaxBuildingLevel();
+
+        if ((citizenHutLevel < citizenHutMaxLevel && Math.pow(2.0, citizenHutLevel + 1.0) <= level) || level >= MAX_CITIZEN_LEVEL)
+        {
+            return;
+        }
+
+        double xpToLevelUp = Math.min(Double.MAX_VALUE, currentXp + xp);
+        while (xpToLevelUp > 0)
```

Comment:
hm can we make this a for loop? kinda don't like that it can get stuck depending on what ExperienceUtils.getXPNeededForNextLevel returns, incase we somehow got to the max level

Reply:
getXPNeededForNextLevel is infinite, even over the max. So that shouldn't be a problem.

Reply:
kk